contracts/cw1-subkeys/src/contract.rs

```diff
@@ -939,6 +1240,88 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_staking_permission_checks() {
+        let mut deps = mock_dependencies(20, &[]);
+
+        let owner = HumanAddr::from("admin0001");
+        let admins = vec![owner.clone()];
+
+        // spender1 has every permission to stake
+        let spender1 = HumanAddr::from("spender0001");
+        // spender2 do not have permission
+        let spender2 = HumanAddr::from("spender0002");
+        let denom = "token1";
+        let amount = 10000;
+        let allow = coin(amount, denom);
+
+        let god_mode = Permissions {
+            delegate: true,
+            redelegate: true,
+            undelegate: true,
+            withdraw: true,
+        };
+
+        let env = mock_env(owner.clone(), &[]);
+        // Init a contract with admins
+        let init_msg = InitMsg {
+            admins: admins.clone(),
+            mutable: true,
+        };
+        init(&mut deps, env.clone(), init_msg).unwrap();
+
+        let setup_perm_msg1 = HandleMsg::SetPermissions {
+            spender: spender1.clone(),
+            permissions: god_mode,
+        };
+        handle(&mut deps, env.clone(), setup_perm_msg1).unwrap();
+
+        let setup_perm_msg2 = HandleMsg::SetPermissions {
+            spender: spender2.clone(),
+            // default is no permission
+            permissions: Default::default(),
+        };
+        // default is no permission
+        handle(&mut deps, env.clone(), setup_perm_msg2).unwrap();
+
+        let msg_delegate = vec![StakingMsg::Delegate {
+            validator: HumanAddr::from("validator1"),
+            amount: allow.clone(),
+        }
+        .into()];
+        let msg_redelegate = vec![StakingMsg::Redelegate {
+            src_validator: HumanAddr::from("validator1"),
+            dst_validator: HumanAddr::from("validator2"),
+            amount: allow.clone(),
+        }
+        .into()];
+        let msg_undelegate = vec![StakingMsg::Undelegate {
+            validator: HumanAddr::from("validator1"),
+            amount: allow.clone(),
+        }
+        .into()];
+        let msg_withdraw = vec![StakingMsg::Withdraw {
+            validator: HumanAddr::from("validator1"),
+            recipient: None,
+        }
+        .into()];
+
+        let msgs = vec![msg_delegate, msg_redelegate, msg_undelegate, msg_withdraw];
+
+        // spender1 can execute
+        for msg in &msgs {
+            let env = mock_env(&spender1, &[]);
+            let res = handle(&mut deps, env, HandleMsg::Execute { msgs: msg.clone() });
+            assert!(res.is_ok())
+        }
+
+        // spender2 cannot execute (no permission)
+        for msg in &msgs {
```

Comment:
I would add a few more tests where one sender has eg. 2 permissions set, 2 unset and ensure that those two pass, and the other 2 fail. 

Just a double check that there is no typo in which permission is checked for which message. And ensures future changes don't break this (what if we swap delegate and redelegate permission checks? these tests would still pass)

Reply:
Added the test. Developer errors must be more distinctive and functional than just a str :( https://github.com/CosmWasm/cosmwasm/issues/518